gpytorch/functions/_diagonalization.py

```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env python3
+
+import torch
+from torch.autograd import Function
+
+from .. import settings
+from ..utils import lanczos
+
+
+class Diagonalization(Function):
+    @staticmethod
+    def forward(ctx, representation_tree, device, dtype, matrix_shape, max_iter, batch_shape, *matrix_args):
```

Comment:
Passing in the lazy tensor instead of the representation tree causes errors in lazy evaluated kernel tensor (which doesn't want to use _matmul without checkpointing).

Reply:
So it's probably best to leave it as it is and just modify when i put it into LinearOperator (after this PR I guess).